### PR TITLE
Save clip to asset

### PR DESCRIPTION
**Pull request by EdgarPironti**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7d221cec0b018e8c958788104f782787821d34ab available at: https://gist.github.com/c91f0061c01428894d28<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6444.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6444.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 50ded685138cff038191065423c86123c42938d0 available at: https://gist.github.com/c91f0061c01428894d28<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6444.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6444.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5be4e0dca8c9e083898d312ef5d4f16a943c5fce available at: https://gist.github.com/313d9fad357ed7af6e9a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6444.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6444.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 54a9e945894a86cfce2bc3cb05b9d32f67bbafae available at: https://gist.github.com/c91f0061c01428894d28<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6444.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6444.dmg

File: libraries/script-engine/src/RecordingScriptingInterface.h [hunk 1]
```diff
@@ -18,6 +18,10 @@
 #include <recording/Forward.h>
 #include <recording/Frame.h>
 
+#include <QtScript/QScriptValue>
+#include <AssetClient.h>
+#include <AssetUpload.h>
```
Comment (Atlante45) [inline libraries/script-engine/src/RecordingScriptingInterface.h]: You don't need those includes in the header, move them to the cpp.
The only thing you need here is to forward declare QScriptValue:
`class QScriptValue;`

File: libraries/script-engine/src/RecordingScriptingInterface.cpp [hunk 2]
```diff
@@ -171,6 +175,45 @@ void RecordingScriptingInterface::saveRecording(const QString& filename) {
     recording::Clip::toFile(filename, _lastClip);
 }
 
+bool RecordingScriptingInterface::saveRecordingToAsset(QScriptValue getClipAtpUrl) {
+	if (!getClipAtpUrl.isFunction())
+        return false;
```
Comment (Atlante45) [inline libraries/script-engine/src/RecordingScriptingInterface.cpp]: The indentation looks wrong here.
Coding standard also requires that you use curly braces even for one line if:
```
if (...) {
    // do stuff
}
```
I think you should probably print a warning so users can figure out why it doesn't work.

File: libraries/script-engine/src/RecordingScriptingInterface.cpp [hunk 3]
```diff
@@ -171,6 +175,45 @@ void RecordingScriptingInterface::saveRecording(const QString& filename) {
...
+
+    if (QThread::currentThread() != thread()) {
+        QMetaObject::invokeMethod(this, "saveRecordingToAsset", Qt::BlockingQueuedConnection,
+            Q_ARG(QScriptValue, getClipAtpUrl));
```
Comment (Atlante45) [inline libraries/script-engine/src/RecordingScriptingInterface.cpp]: You should return the correct value here:
i.e.
```
    if (QThread::currentThread() != thread()) {
        bool result { false };
        QMetaObject::invokeMethod(this, "saveRecordingToAsset", Qt::BlockingQueuedConnection,
                                  Q_ARG(QScriptValue, getClipAtpUrl),
                                  Q_RESULT(bool, result));
        return result;
    }
```

File: libraries/script-engine/src/RecordingScriptingInterface.cpp [hunk 4]
```diff
@@ -171,6 +175,45 @@ void RecordingScriptingInterface::saveRecording(const QString& filename) {
...
+    }
+
+    if (!_lastClip) {
+        qWarning() << "There is no recording to save";
+        return false;
+    }
+
+    if (auto upload = DependencyManager::get<AssetClient>()->createUpload(recording::Clip::toBuffer(_lastClip), HFR_EXTENSION)) {
+        QObject::connect(upload, &AssetUpload::finished, this, [=](AssetUpload* upload, const QString& hash) mutable {
+            auto filename = QFileInfo(upload->getFilename()).fileName();
```
Comment (Atlante45) [inline libraries/script-engine/src/RecordingScriptingInterface.cpp]: Unused

File: libraries/script-engine/src/RecordingScriptingInterface.cpp [hunk 5]
```diff
@@ -171,6 +175,45 @@ void RecordingScriptingInterface::saveRecording(const QString& filename) {
...
+            QString clip_atp_url = "";
+
+            if (upload->getError() == AssetUpload::NoError) {
+
+                clip_atp_url = QString("%1:%2.%3").arg(URL_SCHEME_ATP, hash, upload->getExtension());
+                upload->deleteLater();
+            } else {
+                qDebug() << "Error during the Asset upload.";
```
Comment (Atlante45) [inline libraries/script-engine/src/RecordingScriptingInterface.cpp]: qWarning

File: libraries/script-engine/src/RecordingScriptingInterface.cpp [hunk 6]
```diff
@@ -171,6 +175,45 @@ void RecordingScriptingInterface::saveRecording(const QString& filename) {
...
+
+            QScriptValueList args;
+            args << clip_atp_url;
+            getClipAtpUrl.call(QScriptValue(), args);
+        });
+        upload->start();
+        return true;
+    }
+
+    return false;
```
Comment (Atlante45) [inline libraries/script-engine/src/RecordingScriptingInterface.cpp]: You should probably print a warning here too.

**Comment by Atlante45** [comment]
Thought about it after the fact, but can you switch your `qDebug()` and `qWarning()`
to `qCDebug(scriptengine)` and `qCWarning(scriptengine)` so that they are correctly categorized.

**Comment by hifi-gustavo** [comment]
Build console output for commit: df2153186d6c6005137f0854cd7dad0bddd72c07 available at: https://gist.github.com/c91f0061c01428894d28<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6444.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6444.dmg

**Pull request merged**